packages/expo-image/src/web/AnimationManager.tsx

```diff
@@ -0,0 +1,198 @@
+        );
+      }
+    });
+  };
+
+  return {
+    startingClass: `${animationClass}-start`,
+    run: runAnimation,
+    containerClass: `${animationClass}-container`,
+    timingFunction,
+    animationClass,
+  };
+}
+
+type MountedAnimationNode = {
+  animationKey: string;
+  child: React.ReactElement;
+  ref: React.RefObject<HTMLImageElement>;
+  callbackContainer: {
+    onReady: (() => void) | null;
+    onAnimationFinished: ((forceUnmount?: boolean) => void) | null;
+    onMount: (() => void) | null;
+  };
+};
+
+export default function AnimationManager({
+  children: renderFunction,
+  initial,
+  animation,
+}: {
+  children: AnimationManagerNode;
+  initial: AnimationManagerNode | null;
+  animation: null | Animation;
+}) {
+  const initialNode = useAnimationManagerNode(initial);
+  if (initialNode) {
+    initialNode.callbackContainer.onAnimationFinished = () =>
+      setNodes((n) =>
+        n.filter(
+          (node, index) => node.animationKey !== initialNode.animationKey || index === n.length - 1
+        )
+      );
+  }
+
+  const [nodes, setNodes] = React.useState<MountedAnimationNode[]>(
+    initialNode ? [initialNode] : []
+  );
+  const newNode = useAnimationManagerNode(renderFunction);
+  if (newNode) {
+    newNode.callbackContainer.onAnimationFinished = (forceUnmount = false) => {
+      if (newNode.ref.current?.classList.contains('unmount') || forceUnmount) {
+        setNodes((n) =>
+          n.filter(
+            (node, index) => node.animationKey !== newNode.animationKey || index === n.length - 1
+          )
+        );
+      }
+    };
+  }
+  React.useEffect(() => {
+    setNodes((n) => {
+      if (!newNode) {
+        return n;
+      }
+      const existingNodeIndex = n.findIndex((node) => node.animationKey === newNode.animationKey);
+      if (existingNodeIndex >= 0) {
+        const copy = [...n];
+        copy.splice(existingNodeIndex, 1, newNode);
+        return copy;
+      }
+
+      newNode.callbackContainer.onMount = () => {
+        if (!newNode?.ref.current || !animation?.startingClass) {
+          return;
+        }
+        if (!newNode?.ref.current.classList.contains('transitioning')) {
+          if (animation?.timingFunction) {
+            newNode.ref.current.style.setProperty(
+              '--expo-image-timing',
+              validateTimingFunctionForAnimation(animation.animationClass, animation.timingFunction)
+            );
+          }
+          setClassOnElement(newNode?.ref.current, [animation?.startingClass]);
+        }
+      };
+
+      newNode.callbackContainer.onReady = () => {
+        if (animation) {
+          animation.run(
+            newNode.ref,
+            n.map((n2) => n2.ref)
+          );
+        } else {
+          n.forEach((oldNode) => oldNode.callbackContainer?.onAnimationFinished?.(true));
+        }
+      };
+      n.forEach((prevNode) => (prevNode.callbackContainer.onReady = () => null));
+      return [...n, newNode];
+    });
+  }, [newNode]);
```

Comment:
Shouldn't `animation` be in the dependencies too?

Reply:
Fixed by moving animation elsewhere